pallets/utxo/src/rewards.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) 2021 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://spdx.org/licenses/MIT
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author(s): C. Yap
+
+use crate::{
+    convert_to_h256, BlockAuthor, Config, Event, Pallet, RewardTotal, TransactionOutput, UtxoStore,
+};
+
+use frame_support::traits::Get;
+use sp_runtime::traits::{
+    BlakeTwo256, CheckedDiv, CheckedSub, Hash, One, SaturatedConversion, Zero,
+};
+
+/// handle event when a block author is found.
+impl<T> pallet_authorship::EventHandler<T::AccountId, T::BlockNumber> for Pallet<T>
+where
+    T: Config + pallet_authorship::Config,
+{
+    fn note_author(author: T::AccountId) {
+        match convert_to_h256::<T>(&author) {
+            Ok(author_h256) => {
+                // store the block author. Reward during the `fn finalize()` phase.
+                <BlockAuthor<T>>::put(author_h256);
+            }
+            Err(e) => {
+                log::warn!("failed to find author: {:?}", e);
+            }
+        }
+    }
+
+    fn note_uncle(_author: T::AccountId, _age: T::BlockNumber) {
+        log::info!("TODO: no support for this. Or is there...?");
+    }
+}
+
+/// Returns the newly reduced reward amount for a Block Author.
+/// How much a reward is reduced, is based on the config's`RewardReductionFraction`.
+fn get_reward_amount<T: Config>(block_number: T::BlockNumber) -> u128 {
```

Comment:
This function seems way more complex than what it should be doing. This function shouldn't include fees or anything else, which seems to be the case based on the implementation of the `reward_block_author` function. Please document what exactly is going on in here. I fail to understand how this works and fail to understand why there's a while loop (!) in side a reward calculation function.

Reply:
Couldn't it be something like
` if blocknum < 100 {
    reward = 100; 
  } else if blocknum < 1000 {
    reward = 50; 
  } else {
    reward = 0;
  } `
  
  I mean obviously that's over simplified and messy but the idea holds...

Reply:
@b-yap ?

Reply:
1 year of 1 min blocks is roughly 525600 blocks for example

Reply:
` if blocknum < 525600 { reward = 100; } else if blocknum < (525600 * 2) { reward = 75; }`

Reply:
etc..

Reply:
What Ben mentioned is legitimate. Another thing we can do is just have an array with the rewards, and pick the element based on the block number. Both ways are OK. The only issue with Ben's way is that we have to create multiple functions depending on the consensus rules we want (i.e., testnet, mainnet, etc). If all are the same, that works too. I'm OK with either way.

For me, the most important thing is to have a specific and fixed way to do this that's simple enough to be tested independently in unit tests.

Simplicity, simplicity, then more simplicity. Things that are complex can have mistakes.

Reply:
I'm happy with the array solution too

Reply:
This is different to the mainnet btw. In the mainnet it goes to 0, for the testnet it bottoms out at 1. By choice

Reply:
The array solution will only work if both testnet and mainnet have the same structure. If they don't, we have to find some other solution and two different functions to be called, whether through a trait or something else.

Well, for the mean time, Carla, it seems that we both just hope to see something simple that's testable. Feel free to do whatever you want otherwise.